Allow GnuPG 2 users to export keys with and without passphrase

piterpunk opened the pull request:
### What does this PR do?
Added the `use_passphrase` parameter. This allows to export keys with a passphrase. The passphrase itself comes from pillar `gpg_passphrase`.
Added the `expect_passphrase = False` to Python GnuPG's `export_keys` call when the `use_passphrase` is False (default). That allows the export of passwordless keys.

### What issues does this PR fix or reference?
Fixes: #58980

### Previous Behavior
An error when calling `gpg.export_key`, see the issue #58980
Also, there was no support to use a passphrase.

### New Behavior
It's now possible to export a key with and without a passphrase.

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Commit:
Allow GnuPG 2 users to export keys

- Calls gpg.export_keys with expect_passphrase = False to allow the
  export of keys without passphrase
- Added support to export keys with a passphrase.
- Behaviour controlled by parameter use_passphrase, defaulted to False

piterpunk commented:
All unit tests for gpg execution module are skipped. Some are marked as slow and the others as destructive. I'll try to update the tests soon. But in a clean environment, the "destructive" looks scary.

bryceml commented:
If you can get this working on centos/rhel 8, that would be great, they seem to have tried to remove all methods of programmatically entering passphrases.

It would be great for this to work in conjunction with the pkgbuild.make_repo as well in centos 8.

piterpunk commented:
> If you can get this working on centos/rhel 8, that would be great, they seem to have tried to remove all methods of programmatically entering passphrases. 
> Another issue I've seen is requiring a tty to work, if this could take care of that too, that would be pretty cool.

Let's see what I can do. I still need to write the tests (which usually take a lot of time). Then I will test in other platforms, including CentOS 8.

Commit:
Allow no passphrase to delete_key and create_key

- Added parameter use_passphrase to delete_key
- Added _gen_key_input to send %no-protection command to gen_key

piterpunk commented:
@bryceml ,

Ok, now I have the `gpg` module working Ok for CentOS 8. Needed to change a bit the `create_key` and `delete_key`, in addition of the changes in `export_key` done in the first commit.

Can you check if this covers your usecases?

Still need to update the unit tests.

Commit:
Updated gpg module tests

- Added tests for new _gen_key_input function, used by create_key
- Added tests for export_key
- Updated tests for delete_key

Commit:
Added changelog entry

Commit:
Sorted strings before compare in gpg unit tests

Commit:
Forgot to test one of the tests. Really fixed now

piterpunk commented:
Ok first the tests fails in all Debian derived platforms, then I re-write two tests to fix this but forgot to test one of them, so a new commit to fix the broked test, after that it seems all tests run Ok, except for ubuntu1804/pytest, but it did break in what I guess is an unrelated test in salt-proxy. 

Did a merge with master and finally it pass all tests. Please @xeacott, can you take a look if this PR is Ok while it's all green?

Ch3LL commented on salt/modules/gpg.py:
```diff
@@ -214,6 +215,38 @@ def _search_keys(text, keyserver, user=None):
+    """
+    Generate --gen-key input per gpg doc/DETAILS
+
+    This code is from python-gnupg 0.4.7dev0 it's here because Saltstack
```
If this code is from another project it would need to go into salt/ext, with proper licensing and credit attached to it, but I would prefer not to maintain upstream code unless there is very good reason to. If you need to allow a newer version of gnupg that is okay if its not available by default on all operating systems, because they can still install it from pypi if needed. I would prefer not to include this and instead import the module and use the code from the upstream.

piterpunk replied:
I picked this snippet from the development unreleased code, the important change is the line '%no-protection' before the '%commit'. It seemed easier to do use the code, but it's possible to use the default gen_input_parameters, check the output for the '%no-protection' when the passphrase is empty and edit the output accordingly.

piterpunk replied:
@Ch3LL , I removed the external code and adds the '%no-protection' in the output of `gpg.gen_key_input` call if needed.

The tests are running fine, except for Ubuntu 16.04 which doesn't collect enough entropy to create the keys, but I guess the same can happen with other systems so I'll do an additional commit marking the specific test as slow.

Provided enough entropy the key generation itself is Ok in Ubuntu 16.04:
```
# salt-call grains.get osfinger
local:
    Ubuntu-16.04
# salt-call gpg.create_key use_passphrase=False
local:
    ----------
    fingerprint:
        23037AEB73E710D964E4C113CFA232687258F0FA
    message:
        GPG key pair successfully generated.
    res:
        True
```

piterpunk replied:
@Ch3LL I guess the solicited change was addressed, can you check if it's OK now?

Commit:
Rewrote passphraseless handling in gpg.create_key

- Removed the _gen_key_input function
- Added code to add '%no-protection' independent of python-gnupg version
- Updated the tests to test directly the gpg.create_key

Commit:
Marked some gpg modules tests as slow

The create_key tests are entropy dependent and can wait for a long
time to have enough of it in the automated tests environment.

dwoz commented on salt/modules/gpg.py:
```diff
@@ -493,7 +504,12 @@ def create_key(
+    delete_secret=False,
+    user=None,
+    gnupghome=None,
+    use_passphrase=False,
```
Prior to this change the we would have used the passphrase by default. I think this should still default to true so we don't break existing behavior on people.

piterpunk replied:
Thank you for the reminder. The original behaviour was restored and the tests updated to reflect this.

Commit:
Restore gpg.delete_key default to use_passphrase

s0undt3ch commented on salt/modules/gpg.py:
```diff
@@ -516,6 +532,10 @@ def delete_key(
     gnupghome
         Specify the location where GPG keyring and related files are stored.
 
+    use_passphrase
+        Whether to use a passphrase with the signing key. Passphrase is received
+        from Pillar.
+
```
```suggestion
    use_passphrase
        Whether to use a passphrase with the signing key. Passphrase is received
        from Pillar.
        
        .. versionadded:: 3003
```

s0undt3ch commented on salt/modules/gpg.py:
```diff
@@ -789,6 +827,10 @@ def export_key(keyids=None, secret=False, user=None, gnupghome=None):
     gnupghome
         Specify the location where GPG keyring and related files are stored.
 
+    use_passphrase
+        Whether to use a passphrase with the signing key. Passphrase is received
+        from Pillar.
+
```
```suggestion
    use_passphrase
        Whether to use a passphrase with the signing key. Passphrase is received
        from Pillar.

        .. versionadded:: 3003

```

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
+                            expect_passphrase=False,
+                        )
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
Why do we need to be root?

s0undt3ch replied:
And why is it destructive?
If it messes with tempfiles, and not the actual user's gpg keys, it's not destructive.

And if we're mocking, it's really not destructive, right?

piterpunk replied:
It only touches temporary files and directories. I only followed the trend that was already there with most tests marked as destructive. My bet about the original mark is: if there is a problem in the code handling the `gpghome` setting it will use the user's `gpghome` and will, indeed, be destructive.

The comment about the `root` is the same situation, is inherited from the `delete_key` test that was already there. I didn't understand the reason and leaved it there.

In my local tests I am doing with my user and they goes fine.

> The slow decorator. why are these unit tests marked as slow?

Because they really create the GPG keys and it depends on system's entropy, which is not warranted to be present. It can take eons in a fresh built VM without `haveged` or other entropy generator running.

s0undt3ch replied:
Ok. Let's remove the destructive decorator and leave the slow one, cool?

piterpunk replied:
Ok, I will do that

piterpunk replied:
Removed the destructive decorator

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
                         self.assertEqual(ret, _expected_result)
+                        gnupg_delete_keys.assert_not_called()
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
And why is it destructive?
If it messes with tempfiles, and not the actual user's gpg keys, it's not destructive.

And if we're mocking, it's really not destructive, right?

piterpunk replied:
The same of the previous one.

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
+                            passphrase=GPG_TEST_KEY_PASSPHRASE,
+                        )
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
same here

piterpunk replied:
The same of the previous one.

piterpunk replied:
Removed the destructive decorator

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
+                        ["xxxxxxxxxxxxxxxx"], False, expect_passphrase=False,
+                    )
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
same here

piterpunk replied:
The same of the previous one.

piterpunk replied:
Removed the destructive decorator

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
+                        expect_passphrase=False,
+                    )
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
same here

piterpunk replied:
The same of the previous one.

piterpunk replied:
Removed the destructive decorator

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
+                        assert gpg.export_key("xxxxxxxxxxxxxxxx", use_passphrase=True)
+                    gnupg_export_keys.assert_not_called()
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
same here

piterpunk replied:
The same of the previous one.

piterpunk replied:
Removed the destructive decorator

s0undt3ch commented on tests/unit/modules/test_gpg.py:
```diff
@@ -378,9 +379,410 @@ def test_delete_key(self):
+                        ["xxxxxxxxxxxxxxxx"], False, passphrase=GPG_TEST_KEY_PASSPHRASE,
+                    )
+
+    @pytest.mark.destructive_test  # Need to run as root!?
```
same here

piterpunk replied:
The same of the previous one.

piterpunk replied:
Removed the destructive decorator

s0undt3ch reviewed (changes requested):
All the tests that were decorated as destructive, if they change temp files or paths, they are not destructive, temp resources are meant to be "destroyable".

The slow decorator. why are these unit tests marked as slow?

s0undt3ch commented on salt/modules/gpg.py:
```diff
@@ -516,6 +532,11 @@ def delete_key(
     gnupghome
         Specify the location where GPG keyring and related files are stored.
 
+    use_passphrase
+        Whether to use a passphrase with the signing key. Passphrase is received
+        from Pillar.
+        
+        .. versionadded:: 3003
```
```suggestion
    use_passphrase
    
        .. versionadded:: 3003
        
        Whether to use a passphrase with the signing key. Passphrase is received
        from Pillar.
```

piterpunk replied:
The initial problem were spaces in the empty line before `.. versionadded`, so it fails with trailing spaces errors. Fixed now.

s0undt3ch commented on salt/modules/gpg.py:
```diff
@@ -789,6 +828,12 @@ def export_key(keyids=None, secret=False, user=None, gnupghome=None):
     gnupghome
         Specify the location where GPG keyring and related files are stored.
 
+    use_passphrase
+        Whether to use a passphrase with the signing key. Passphrase is received
+        from Pillar.
+
+        .. versionadded:: 3003
+
```
```suggestion
    use_passphrase

        .. versionadded:: 3003
        
        Whether to use a passphrase with the signing key. Passphrase is received
        from Pillar.

```

piterpunk replied:
The initial problem were spaces in the empty line before `.. versionadded`, so it fails with trailing spaces errors. Fixed now.

s0undt3ch reviewed (changes requested):
I think this way the docs will build properly

s0undt3ch commented:
And lets also remove the destructive decorator

piterpunk commented:
> And lets also remove the destructive decorator

Ok, by night I can do the changes, test locally and then update this feature branch.

Commit:
Minor adjusts in gpg module and its tests

- Added versionadded information to use_passphrase argument in
  delete_key and export_key
- Unmarked the tests that only writes at temporary files as destructive

piterpunk commented:
@s0undt3ch , just did the required changes. I hope it's now Ok.

Ch3LL committed:
Merge branch 'master' into export-gpg-key-with-and-without-passphrase_Fix-Issue-58980
